x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/support/KerberosTicketValidator.java

```diff
@@ -0,0 +1,254 @@
+        }
+    }
+
+    /**
+     * Privileged wrapper for closing GSSContext, does not throw exceptions but logs
+     * them as warning.
+     *
+     * @param gssContext
+     */
+    private static void privilegedCloseNoThrow(final GSSContext gssContext) {
+        if (gssContext != null) {
+            try {
+                AccessController.doPrivileged((PrivilegedExceptionAction<Void>) () -> {
+                    gssContext.dispose();
+                    return null;
+                });
+            } catch (PrivilegedActionException e) {
+                RuntimeException rte = ExceptionsHelper.convertToRuntime((Exception) ExceptionsHelper.unwrapCause(e));
+                LOGGER.debug("Could not dispose GSS Context", rte);
+            }
+            return;
+        }
+    }
+
+    /**
+     * Privileged wrapper for closing LoginContext, does not throw exceptions but
+     * logs them as warning.
+     *
+     * @param loginContext
+     */
+    private static void privilegedDisposeNoThrow(final LoginContext loginContext) {
+        if (loginContext != null) {
+            try {
+                AccessController.doPrivileged((PrivilegedExceptionAction<Void>) () -> {
+                    loginContext.logout();
+                    return null;
+                });
+            } catch (PrivilegedActionException e) {
+                RuntimeException rte = ExceptionsHelper.convertToRuntime((Exception) ExceptionsHelper.unwrapCause(e));
+                LOGGER.debug("Could not close LoginContext", rte);
+            }
+        }
+    }
+
+    /**
+     * Performs authentication using provided principal name and keytab
+     *
+     * @param principal Principal name
+     * @param keytabFilePath Keytab file path
+     * @param settings {@link Settings}
+     * @return authenticated {@link LoginContext} instance. Note: This needs to be
+     *         closed {@link LoginContext#logout()} after usage.
+     * @throws PrivilegedActionException
+     */
+    private static LoginContext serviceLogin(final String principal, final String keytabFilePath, final Settings settings)
+            throws PrivilegedActionException {
+        return AccessController.doPrivileged((PrivilegedExceptionAction<LoginContext>) () -> {
+            final Set<Principal> principals = new HashSet<>();
+            if (principal.equals("*") == false) {
+                principals.add(new KerberosPrincipal(principal));
+            }
+
+            final Subject subject = new Subject(false, principals, new HashSet<Object>(), new HashSet<Object>());
+
+            final Configuration conf = new KeytabJaasConf(principal, keytabFilePath, settings);
+            final LoginContext loginContext = new LoginContext(KEY_TAB_CONF_NAME, subject, null, conf);
+            loginContext.login();
+            return loginContext;
+        });
+    }
+
+    /**
+     * Instead of jaas.conf, this requires refresh of {@link Configuration}.
```

Comment:
add a complete description. This is just a fragment that doesn't make sense to those trying to understand the code.